src/riak_core_throttle.erl

```diff
@@ -0,0 +1,301 @@
+%% to represent numerically the amount of load that an activity is under and
+%% map that level of load to a particular throttle level. Examples of metrics
+%% that could be used for load factors include the size of the mailbox for a
+%% particular process, or the number of messages per second being sent to an
+%% external service such as Solr.
+-module(riak_core_throttle).
+
+%% API
+-export([get_throttle/1,
+         set_throttle/2,
+         clear_throttle/1,
+         disable_throttle/1,
+         enable_throttle/1,
+         is_throttle_enabled/1,
+         init/4,
+         set_limits/2,
+         get_limits/1,
+         clear_limits/1,
+         create_limits_translator_fun/2,
+         set_throttle_by_load/2,
+         throttle/1]).
+
+-ifdef(TEST).
+-export([get_throttle_for_load/2]).
+-define(SLEEP(Time), Time).
+-else.
+-define(SLEEP(Time), timer:sleep(Time)).
+-endif.
+
+-type activity_key() :: atom().
+-type throttle_time() :: non_neg_integer().
+-type load_factor() :: number() | atom() | tuple().
+-type limits() :: [{load_factor(), throttle_time()}].
+
+-export_type([activity_key/0, throttle_time/0, load_factor/0, limits/0]).
+
+-define(THROTTLE_KEY(Key), list_to_atom("throttle_" ++ atom_to_list(Key))).
+-define(THROTTLE_LIMITS_KEY(Key),
+        list_to_atom("throttle_limits_" ++ atom_to_list(Key))).
+-define(THROTTLE_KILL_KEY(Key),
+        list_to_atom("throttle_kill_" ++ atom_to_list(Key))).
+
+%% @doc Sets the throttle for the activity identified by `Key' to the specified
+%% `Time'.
+-spec set_throttle(activity_key(), throttle_time()) -> ok.
+set_throttle(Key, Time) when Time >= 0 ->
+    application:set_env(riak_core, ?THROTTLE_KEY(Key), Time).
+
+%% @doc Clears the throttle for the activity identified by `Key'.
+-spec clear_throttle(activity_key()) -> ok.
+clear_throttle(Key) ->
+    application:unset_env(riak_core, ?THROTTLE_KEY(Key)).
+
+%% @doc Disables the throttle for the activity identified by `Key'.
+-spec disable_throttle(activity_key()) -> ok.
+disable_throttle(Key) ->
+    application:set_env(riak_core, ?THROTTLE_KILL_KEY(Key), true).
+
+%% @doc Enables the throttle for the activity identified by `Key'.
+-spec enable_throttle(activity_key()) -> ok.
+enable_throttle(Key) ->
+    application:unset_env(riak_core, ?THROTTLE_KILL_KEY(Key)).
+
+%% @doc Returns `true' if the throttle for the activity identified by `Key' is
+%% enabled, `false' otherwise.
+-spec is_throttle_enabled(activity_key()) -> boolean().
+is_throttle_enabled(Key) ->
+    Killed = application:get_env(riak_core, ?THROTTLE_KILL_KEY(Key), false),
+    not Killed.
+
+%% @doc Initializes the throttle for the activity identified by `Key' using
+%% configuration values found in the application environment for `AppName'.
+-spec init(Key::activity_key(),
+           AppName::atom(),
+           {LimitsKey::atom(), LimitsDefault::limits()},
+           {KillSwitchKey::atom(), KillSwitchDefault::boolean()}) ->
+    ok | {error, Reason :: term()}.
+init(Key,
```

Comment:
Switch order of Key and AppName as we're going to pass AppName everywhere else, and it has really become a 2-level hierarchy.